### PR TITLE
enh(gdrive) auto-skip spreadsheet file if too many internal errors

### PR DESCRIPTION
**Pull request by fontanierh**
## Description

Sometimes, a google drive spreadsheet will consistently return a 500 internal error from Google.
In these cases, the only option is to skip the file.

This commit does just that: add some in-activity retries (3) on internal errors (confirming it's not a fluke) and automatically skip that file if the error persists, after 20 activity failures. This would make it highly unlikely that we'd skip a file for a transient internal error (which is also a thing).

### CODE REVIEW

File: connectors/src/connectors/google_drive/temporal/spreadsheets.ts [hunk 1]
```diff
@@ -424,7 +425,36 @@ export async function syncSpreadSheet(
...
+  let internalErrorsCount = 0;
+  const maxInternalErrors = 3;
+  let spreadsheet: Awaited<ReturnType<typeof getSpreadsheet>>;
+  // If we consistently get 500 Internal Server Error from Google Sheets after 20 activity failures, we skip the file.
+  for (;;) {
+    try {
+      spreadsheet = await getSpreadsheet();
+      break;
+    } catch (err) {
+      if (err instanceof Error && "code" in err && err.code === 500) {
```
Comment (flvndvd) [inline connectors/src/connectors/google_drive/temporal/spreadsheets.ts:437]: Can we move this logic to `getSpreadsheet` so all the logic remains close to the call site?

File: connectors/src/connectors/google_drive/temporal/spreadsheets.ts [hunk 2]
```diff
@@ -424,7 +425,36 @@ export async function syncSpreadSheet(
     }
   };
 
-  const spreadsheet = await getSpreadsheet();
+  let internalErrorsCount = 0;
+  const maxInternalErrors = 3;
```
Comment (flvndvd) [inline connectors/src/connectors/google_drive/temporal/spreadsheets.ts:429]: You mention 20, but here it says 3, am I missing something?

File: connectors/src/connectors/google_drive/temporal/spreadsheets.ts (hunk 1 above)
Comment (fontanierh) [reply connectors/src/connectors/google_drive/temporal/spreadsheets.ts:437]: How about I get rid of getSpreadsheet entirely ? That way I can directly return from the `catch`

Comment (fontanierh) [reply connectors/src/connectors/google_drive/temporal/spreadsheets.ts:437]: Haha you hate those do you

File: connectors/src/connectors/google_drive/temporal/spreadsheets.ts [hunk 3]
```diff
@@ -420,39 +427,25 @@ export async function syncSpreadSheet(
...
-  const maxInternalErrors = 3;
-  let spreadsheet: Awaited<ReturnType<typeof getSpreadsheet>>;
-  // If we consistently get 500 Internal Server Error from Google Sheets after 20 activity failures, we skip the file.
-  for (;;) {
-    try {
-      spreadsheet = await getSpreadsheet();
-      break;
-    } catch (err) {
-      if (err instanceof Error && "code" in err && err.code === 500) {
+      } else if (err instanceof Error && "code" in err && err.code === 500) {
```
Comment (flvndvd) [inline connectors/src/connectors/google_drive/temporal/spreadsheets.ts:430]: `nit` could be move to a typeguard like above.

**Review by flvndvd — APPROVED** [review]
LGTM, hopefully, this agressive retry does not hit the rate limit.

File: connectors/src/connectors/google_drive/temporal/spreadsheets.ts (hunk 3 above)
Comment (fontanierh) [reply connectors/src/connectors/google_drive/temporal/spreadsheets.ts:430]: good point, will move to it in a separate PR, but have limited appetite for it right now (just need to unstuck)

**Pull request merged**